components/graph/git-elements.ts

```diff
@@ -0,0 +1,629 @@
+    this.isLocalTag = this.name.indexOf('tag: ') === 0;
+    this.isTag = this.isLocalTag || this.isRemoteTag;
+    const isRemoteBranchOrHEAD = this.name.indexOf('refs/remotes/') == 0;
+    this.isLocalHEAD = this.name == 'HEAD';
+    this.isRemoteHEAD = this.name.includes('/HEAD');
+    this.isLocalBranch = this.name.indexOf('refs/heads/') == 0;
+    this.isRemoteBranch = isRemoteBranchOrHEAD && !this.isRemoteHEAD;
+    this.isStash = this.name.indexOf('refs/stash') == 0;
+    this.isHEAD = this.isLocalHEAD || this.isRemoteHEAD;
+    this.isBranch = this.isLocalBranch || this.isRemoteBranch;
+    this.isRemote = isRemoteBranchOrHEAD || this.isRemoteTag;
+    this.isLocal = this.isLocalBranch || this.isLocalTag;
+    if (this.isLocalBranch) {
+      this.localRefName = this.name.slice('refs/heads/'.length);
+      this.refName = this.localRefName;
+    }
+    if (this.isRemoteBranch) {
+      this.localRefName = this.name.slice('refs/remotes/'.length);
+    }
+    if (this.isLocalTag) {
+      this.localRefName = this.name.slice('tag: refs/tags/'.length);
+      this.refName = this.localRefName;
+    }
+    if (this.isRemoteTag) {
+      this.localRefName = this.name.slice('remote-tag: '.length);
+    }
+    const splitedName = (this.localRefName || this.name).split('/');
+    if (this.isRemote) {
+      // get rid of the origin/ part of origin/branchname
+      this.remote = splitedName[0];
+      this.refName = splitedName.slice(1).join('/');
+    }
+    this.show = true;
+    this.server = this.graph.server;
+    this.isDragging = ko.observable(false);
+    this.current = ko.pureComputed(
+      () => this.isLocalBranch && this.graph.checkedOutBranch() == this.refName
+    );
+    this.color = this._colorFromHashOfString(this.name);
+
+    this.node.subscribe(
+      (oldNode) => {
+        if (oldNode) oldNode.removeRef(this);
+      },
+      null,
+      'beforeChange'
+    );
+    this.node.subscribe((newNode) => {
+      if (newNode) newNode.pushRef(this);
+    });
+
+    // This optimization is for autocomplete display
+    this.value = splitedName[splitedName.length - 1];
+  }
+
+  displayHtml(largeCurrent) {
+    const size = largeCurrent && this.current() ? 26 : 18;
+    let prefix = '';
+    if (this.isRemote) {
+      prefix = `<span>${octicons.globe.toSVG({ height: size })}</span> `;
+    }
+    if (this.isBranch) {
+      prefix += `<span>${octicons['git-branch'].toSVG({ height: size })}</span> `;
+    } else if (this.isTag) {
+      prefix += `<span>${octicons.tag.toSVG({ height: size })}</span> `;
+    }
+    return prefix + this.localRefName;
+  }
+
+  _colorFromHashOfString(string) {
+    return `#${md5(string).toString().slice(0, 6)}`;
+  }
+
+  dragStart() {
+    ungit._perf.dd.push(5);
+    this.graph.currentActionContext(this);
+    this.isDragging(true);
+    if (document.activeElement) {
+      (document.activeElement as HTMLElement).blur();
+    }
+  }
+
+  dragEnd() {
+    ungit._perf.ee.push(6);
```

Comment:
What is this for?

Reply:
Very crude way of performance monitoring... haha it's removed.